Add testing functions (expect_linear/loglinear/quadratic_time) + testthat files

Anirban166 opened the pull request:
- Added simple testing functions for the three main complexity classes (linear, log-linear, quadratic) following whats written [here](https://github.com/rstats-gsoc/gsoc2020/wiki/Asymptotic-complexity-testing): 
> "testing functions, using syntax similar to testthat package. For example expect_linear_time(fun(N)) should fail with an error if fun(N) is not linear in N, i.e. O(N)."

- Added testthat files such as one for constrained changepoint model algorithms (more functions will be added later) with initial test cases of `PeakSegOptimal::PeakSegPDPA` and `PeakSegDP::cDPA` made with the help of the above functions (syntax similar to testthat) and placed the optional packages in suggests as mentioned [here](https://github.com/Anirban166/testComplexity/issues/5)

@tdhock are these changes okay? Also are the testing functions in the form you expected them to be? please review once

tdhock commented on R/asymptoticTimings.R:
```diff
@@ -32,15 +32,19 @@ asymptoticTimings <- function(e, data.sizes, max.seconds)
     timings.list[[i]] <- data.frame(benchmarked.timings$time, benchmarked.timings$data.size)
 
-    if(mean(benchmarked.timings$time) > time.limit) break else next
+    if(mean(benchmarked.timings$time) > time.limit) break
```
again please move this condition into the while(), please avoid using break, it is confusing

Anirban166 replied:
sure, done

tdhock commented on R/expect_linear_time.R:
```diff
@@ -0,0 +1,18 @@
+{
+  timings.df <- f(...)
+
+  stopifnot(asymptoticTimeComplexityClass(timings.df) == "linear")
```
stopifnot does not provide an informative error message so please use stop with an informative error message instead.

Anirban166 replied:
yup right, I thought about using `stop(msg)` for the same reason but I forgot to change it!

tdhock commented on R/expect_loglinear_time.R:
```diff
@@ -0,0 +1,18 @@
+{
+  timings.df <- f(...)
+
+  stopifnot(asymptoticTimeComplexityClass(timings.df) == "loglinear")
```
since the only difference between this function and the other expect_*_time functions is the complexity class, you should make a generic funciton like 
```
expect_time_complexity <- function(complexity.class, f){......}
```
and then use it from the other expect functions like
```
expect_loglinear_time <- function(...)expect_time_complexity("loglinear", ...)
```
  expect_time_complexity(..., c

Anirban166 replied:
If I get you right then it would mean `expect_time_complexity` would contain the complexity comparison part (to avoid that duplication) and it would be internally called in each of the different complexity classes' expect functions with the corresponding class, right?  

But then wouldn't a single function with the added `complexity.class` parameter work?
```r
expect_time_complexity = function(complexity.class, f, ...)
{
  timings.df <- f(...)
  if(asymptoticTimeComplexityClass(timings.df) != complexity.class)
    stop("Complexity mismatch: Expected ", complexity.class, " complexity")
}
```
Also would it be better to return the string (can use `invisible()` in conjunction as well) denoting the complexity class in the complexity classifying functions instead of printing it? It doesn't seem to provide any issue, but it stacks up, as visible in a RCMD `check()`.  <br>
Can also use:
```r
expect_time_complexity = function(expected.complexity.class, f, ...)
{
  timings.df <- f(...)
  predicted.complexity.class <- asymptoticTimeComplexityClass(timings.df)
  if(predicted.complexity.class != expected.complexity.class)
    stop("Complexity mismatch: Expected ", expected.complexity.class, " complexity, in place of the predicted ", predicted.complexity.class, " complexity")
}
```

tdhock replied:
actually instead of stop you should use expect_identical or some other expect_ function (maybe write your own?) which gives an informative error message.

Anirban166 replied:
btw what should the error message contain to be informative? there aren't any tracebacks when I use `expect_xyz` functions, so am wondering is it just a casual message regarding string mismatch, or something which a `try()` produces?

Anirban166 commented:
- Added `expect_time_complexity`, which will take care of all expected complexity testing cases, example:
```r
expect_time_complexity("quadratic", PeakSegDP::cDPA(rpois(data.sizes, 1), rep(1, length(rpois(data.sizes, 1))), 3L), data.sizes = 10^seq(1, 4, by = 0.5))
[1] quadratic
```
- Updated `expect_x_time` functions (x = complexity class) to call `expect_time_complexity` internally, with the only requirement as a parameter being the ones passed onto `asymptoticTimings` (no need to pass it, and complexity class will be defined by x) Example usage:
```r
expect_quadratic_time(PeakSegDP::cDPA(rpois(data.sizes, 1), rep(1, length(rpois(data.sizes, 1))), 3L), data.sizes = 10^seq(1, 4, by = 0.5))
[1] quadratic
```
